python/libsettings.pyx

```diff
@@ -0,0 +1,310 @@
+    #                                     bytes(section),
+    #                                     bytes(name),
+    #                                     my_int,
+    #                                     cython.sizeof(int),
+    #                                     stype,
+    #                                     my_int_notify,
+    #                                     my_int)
+
+    def write(self, section, name, value):
+        return settings_write_str(self.ctx,
+                                  bytes(section),
+                                  bytes(name),
+                                  bytes(str(value)))
+
+    def read(self, section, name):
+        cdef char value[256]
+        ret = settings_read_str(self.ctx,
+                                bytes(section),
+                                bytes(name),
+                                value,
+                                256)
+        if (ret):
+            return None
+        else:
+            return str(value)
+
+    def read_all(self):
+        cdef char section[256]
+        cdef char name[256]
+        cdef char value[256]
+        cdef char fmt_type[256]
+        cdef uint16_t idx = 0
+
+        l = []
+
+        while (True):
+            ret = settings_read_by_idx(self.ctx,
+                                       idx,
+                                       section,
+                                       256,
+                                       name,
+                                       256,
+                                       value,
+                                       256,
+                                       fmt_type,
+                                       256)
+
+            if (ret > 0):
+                break
+            elif (ret < 0):
+                return []
+
+            s = {
+                    'section': str(section),
+                    'name': str(name),
+                    'value': str(value),
+                    'fmt_type': str(fmt_type),
+                }
+
+            l.append(s)
+            idx += 1
+
+        return l
+
+    def _callback_broker(self, sbp_msg, **metadata):
+        if self._debug:
+            print '_callback_broker', sbp_msg.msg_type
+            print ":".join("{:02x}".format(ord(c)) for c in sbp_msg.payload)
+
+        msg_type = sbp_msg.msg_type
+
+        for k,v in self._callbacks.items():
+            if v[0] == msg_type:
+                cb_data = self._callbacks[k]
+
+        cdef sbp_msg_callback_t f = <sbp_msg_callback_t><uintptr_t>cb_data[1]
+        f(sbp_msg.sender, sbp_msg.length, bytes(sbp_msg.payload), <void*><uintptr_t>cb_data[2])
+
+cdef int send_wrapper(void *ctx, uint16_t msg_type, uint8_t length, uint8_t *payload):
+    settings = <object>ctx
+    
+    if settings._debug:
+        print 'send_wrapper', msg_type
+        print ":".join("{:02x}".format(ord(c)) for c in payload[:length])
+
+    # https://cython.readthedocs.io/en/latest/src/tutorial/strings.html
```

Comment:
Removed.

Reply:
I'll take it back. Copying is necessary. See `Passing byte strings` section in the link:

Alternative was tested not to work:
payload=<bytes>payload

```
This creates a Python byte string object that holds a copy of the original C string. It can be safely passed around in Python code, and will be garbage collected when the last reference to it goes out of scope. It is important to remember that null bytes in the string act as terminator character, as generally known from C. The above will therefore only work correctly for C strings that do not contain null bytes.

Besides not working for null bytes, the above is also very inefficient for long strings, since Cython has to call strlen() on the C string first to find out the length by counting the bytes up to the terminating null byte. In many cases, the user code will know the length already, e.g. because a C function returned it. In this case, it is much more efficient to tell Cython the exact number of bytes by slicing the C string
```

The crucial part being `The above will therefore only work correctly for C strings that do not contain null bytes` and our settings are notorious for having null bytes.

Reply:
Ok, I think explicitly calling out `The above will therefore only work correctly for C strings that do not contain null bytes` would be useful in a comment, but only if you're going to make another rev of the code.